okio/src/main/java/okio/Options.java

```diff
@@ -16,18 +16,233 @@
 package okio;
 
 import java.util.AbstractList;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
 import java.util.RandomAccess;
 
 /** An indexed set of values that may be read with {@link BufferedSource#select}. */
 public final class Options extends AbstractList<ByteString> implements RandomAccess {
   final ByteString[] byteStrings;
+  final int[] trie;
 
-  private Options(ByteString[] byteStrings) {
+  private Options(ByteString[] byteStrings, int[] trie) {
     this.byteStrings = byteStrings;
+    this.trie = trie;
   }
 
   public static Options of(ByteString... byteStrings) {
-    return new Options(byteStrings.clone()); // Defensive copy.
+    if (byteStrings.length == 0) throw new IllegalArgumentException("no options provided");
+
+    // Sort the byte strings which is required when recursively building the trie. Map the sorted
+    // indexes to the caller's indexes.
+    List<ByteString> list = new ArrayList<>(Arrays.asList(byteStrings));
+    Collections.sort(list);
+    List<Integer> indexes = new ArrayList<>();
+    for (int i = 0; i < list.size(); i++) {
+      indexes.add(-1);
+    }
+    for (int i = 0; i < list.size(); i++) {
+      int sortedIndex = Collections.binarySearch(list, byteStrings[i]);
+      indexes.set(sortedIndex, i);
+    }
+    if (list.get(0).size() == 0) {
+      throw new IllegalArgumentException("the empty byte string is not a supported option");
+    }
+
+    // Strip elements that will never be returned because they follow their own prefixes. For
+    // example, if the caller provides ["abc", "abcde"] we will never return "abcde" because we
+    // return as soon as we encounter "abc".
+    for (int a = 0; a < list.size(); a++) {
+      ByteString prefix = list.get(a);
+      for (int b = a + 1; b < list.size(); ) {
+        ByteString byteString = list.get(b);
+        if (!byteString.startsWith(prefix)) break;
+        if (byteString.size() == prefix.size()) {
+          throw new IllegalArgumentException("duplicate option: " + byteString);
+        }
+        if (indexes.get(b) > indexes.get(a)) {
+          list.remove(b);
```

Comment:
We're allocating a `List` and copying the input every time, however this branch may not be reached at all. Would it be better to stick to arrays, mark an element if it needs to be removed, and copy elements into a smaller array if anything has been marked?

Reply:
The logic gets awkward if there are multiple such elements to remove. I’m more annoyed that we allocate a bunch of boxed integers!